Detect deltaMode - fixes IE11/FF49/Chrome54 on Win7, see #9526

fredludlow opened the pull request:
See #9526

fredludlow committed:
Detect deltaMode - fixes IE11/FF49/Chrome54 on Win7, see #9526

mrdoob commented on examples/js/controls/TrackballControls.js:
```diff
@@ -473,7 +473,7 @@ THREE.TrackballControls = function ( object, domElement ) {
 		event.stopPropagation();
 
-		_zoomStart.y -= event.deltaY * 0.01;
+		_zoomStart.y -= event.deltaY * ( ( event.deltaMode )  ? 0.01 : 0.0001 );
```
One could think that `event.deltaMode` is a boolean. How about...

```js
( event.deltaMode === 0 ? 0.0001 : 0.01 )
```

fredludlow commented:
Or even a  switch  case?  (value of 2 means unit is pages). Let me know which of those you prefer and I'll update

mrdoob commented:
Hmm... In which case would `event.deltaMode` be `2`?

mrdoob commented:
Yeah... A switch would be the safest indeed. So we can ignore `2`.

fredludlow commented:
deltaMode takes 0 for pixels, 1 for lines and 2 for pages
https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaMode

Without knowing of a platform that produces a value of 2 it's hard to test but we could assume a page is about 30 lines (approx order of magnitude) and scale accordingly?

mrdoob commented:
I think I would ignore `2` for now.

fredludlow committed:
Add switch for mousewheel handling on wheel event.deltaMode

fredludlow commented:
Okay, I've updated it with a switch, can handle case `2` if anyone ever finds a platform that generates it! 

Also changed the scaling for case `0` to 0.00025 so that FF, Chrome and IE feel more consistent.

MrOpposite commented:
I've successfully managed to get a WheelEvent with deltaMode: 0 on chrome in windows 10.

This is the result of my tests (i simply did `window.addEventListener("wheel",(e)=>console.log(e))` in this browser-window)
![image](https://cloud.githubusercontent.com/assets/1394292/20324950/e35de4fa-ab82-11e6-85ad-98837a381371.png)

This is the setting I changed in windows. Notice the "Roll the mousewheel to scroll" being set to "One screen at a time"
![image](https://cloud.githubusercontent.com/assets/1394292/20324956/e975a38c-ab82-11e6-969d-f4c218d6cc97.png)

I'm unsure what needs testing on this issue though, so I couldn't provide anything in that context.

fredludlow commented:
Thanks @MrOpposite. So... if `deltaMode === 2` then `deltaY` is approximately 1/-1. For `deltaMode=== 0` `deltaY ~= 100`. So I'll update with a `case 2` and scale by 0.025..?

MrOpposite commented:
Are we sure we want to use the deltaY as a scalefactor for zoom at all? Since the deltaY value is defined by how much a user wants a page to scroll up/down, rather than how much to zoom in an application like this. 

I'd actually advocate ignoring the deltaMode entirely and just use `Math.sign(deltaY)` in order to get the direction and dictate the sensitivity ourselves entirely without relying on sane settings of the client.

mrdoob commented:
> I'd actually advocate ignoring the deltaMode entirely and just use `Math.sign(deltaY)` in order to get the direction and dictate the sensitivity ourselves entirely without relying on sane settings of the clients OS.

We lose smooth scrolling (smooth zooming) if we do that...

MrOpposite commented:
Ah, yeah, that's a very good reason to keep battle this issue while still using the value of deltaY...

fredludlow committed:
Now handling deltaMode 2

fredludlow commented:
I've updated my branch to what I suggest we do for now (handle all three modes and scale accordingly), at least until we get another report of a system this doesn't work on. Very open to suggestions/instructions to do it another way if preferred, @mrdoob?

mrdoob commented:
LGTM!

mrdoob commented:
Thanks!
